crates/primitives/src/transaction/mod.rs

```diff
@@ -372,6 +373,22 @@ impl Transaction {
         }
     }
 
+    /// Returns the nonce that was actually used as part of transaction execution
+    /// Returns None if the effective nonce is not known
+    #[cfg(feature = "optimism")]
+    pub fn effective_nonce(&self) -> Option<u64> {
+        match self {
+            Transaction::Legacy(TxLegacy { nonce, .. }) => Some(*nonce),
+            Transaction::Eip2930(TxEip2930 { nonce, .. }) => Some(*nonce),
+            Transaction::Eip1559(TxEip1559 { nonce, .. }) => Some(*nonce),
+            Transaction::Eip4844(TxEip4844 { nonce, .. }) => Some(*nonce),
+            // In deposit transactions, nonces are optional. They are never included in the RLP
+            // encoded deposit transaction even if they are present.
+            #[cfg(feature = "optimism")]
+            Transaction::Deposit(TxDeposit { nonce, .. }) => *nonce,
+        }
```

Comment:
This should make it easier for the upstream maintenance in case the nonce function changes for any of these transaction types.
```suggestion
        match self {
            // In deposit transactions, nonces are optional. They are never included in the RLP
            // encoded deposit transaction even if they are present.
            #[cfg(feature = "optimism")]
            Transaction::Deposit(TxDeposit { nonce, .. }) => *nonce,
            _ => Some(self.nonce()),
```